greenpeace/planet4-master-theme
# PLANET-5727: Limit of post revisions

Ref: [PLANET-5727](https://jira.greenpeace.org/browse/PLANET-5727)

## Description
We should define a limit of post `revisions` and avoid recording unnecessary records into the database. Currently, we don't have a limit for that which means that every time the editor save the post it's store a new revision.

I've reproduced the same as the ticket description, using the international site database and of course with different dates, and got worst results.

#### Using the planet4_international database as a part of this experiment I get (March 9, 2023)
Total size: **359,3 MB**
wp_posts table (with revisions): **272,8 MB**
wp_posts table (without revisions): **89,3 MB**
New size: ~**175,8 MB** which represents ~**51%** of size removed.

#### Example of query to get all revisions grouped by post
```sql
SELECT count(ID) AS total_revisions, ID, post_parent FROM wp_posts WHERE post_type = "revision" GROUP BY post_parent ORDER BY post_date ASC;
+-----------------+-------+-------------+
| total_revisions | ID    | post_parent |
+-----------------+-------+-------------+
|               1 | 61952 |       11442 |
|               1 | 60932 |       11452 |
|               1 | 61951 |        1320 |
|               1 | 61961 |       18532 |
|               1 | 61958 |       54800 |
|               1 | 61956 |       49031 |
|               1 | 61954 |       21155 |
|               1 | 61955 |       25526 |
|               1 | 61962 |       58548 |
|               1 | 60936 |       60796 |
|               1 | 61960 |       61511 |
|               3 | 45241 |       45227 |
|               3 | 45260 |       45226 |
|               3 | 45271 |       45269 |
|               2 | 45280 |       45278 |
|               3 | 61965 |       20416 |
|               3 | 61984 |       56556 |
|               3 | 61989 |       57850 |
|               3 | 61993 |       61276 |
|               3 | 62006 |       58980 |
|               3 | 62050 |        9129 |
|               3 | 62058 |       54446 |
|               3 | 62068 |       45080 |
|               3 | 62071 |       54964 |
|               1 | 61950 |       61949 |
|               3 | 62110 |       45224 |
+-----------------+-------+-------------+
```

#### Remove all `revisions`
```sql
mysql> select count(ID) AS total_revisions from wp_posts where post_type = 'revision';
+----------+
| total_revisions |
+----------+
|    19812 |
+----------+
```

```sql
mysql> delete from wp_posts where post_type = 'revision';
Query OK, 19812 rows affected (7.11 sec)
```

```sql
mysql> select count(ID) as total_revisions from wp_posts where post_type = 'revision';
+----------+
| total_revisions |
+----------+
|        0 |
+----------+
```
**Other queries.**
Total of revisions since the last week
```sql
mysql> SELECT COUNT(ID) as total_revisions FROM wp_posts AS posts WHERE posts.post_type = "revision" AND posts.post_modified BETWEEN date_sub(now(), INTERVAL 1 WEEK) AND now();
+----------+
| total_revisions |
+----------+
|        0 |
+----------+
```

Total of revisions since the last 2 weeks
```sql
mysql> SELECT COUNT(ID) AS total_revisions  FROM wp_posts AS posts WHERE posts.post_type = "revision" AND posts.post_modified BETWEEN date_sub(now(), INTERVAL 2 WEEK) AND now();
+----------+
| COUNT(*) |
+----------+
|       23 |
+----------+
```

Total of revisions since the last 3 weeks
```sql
mysql> SELECT COUNT(ID) AS total_revisions FROM wp_posts AS posts WHERE posts.post_type = "revision" AND posts.post_modified BETWEEN date_sub(now(), INTERVAL 3 WEEK) AND now();
+----------+
| total_revisions |
+----------+
|       74 |
+----------+
```

Total of revisions since the last 4 weeks
```sql
mysql> SELECT COUNT(*) FROM wp_posts AS posts WHERE posts.post_type = "revision" AND posts.post_modified BETWEEN date_sub(now(), INTERVAL 4 WEEK) AND now();
+----------+
| COUNT(*) |
+----------+
|      143 |
+----------+
```

I've also added an option, under `Settings > Writing`, which allows to editors to change the desired revisions in case they prefer more or less to keep.

![Screenshot 2023-03-10 at 10 30 43](https://user-images.githubusercontent.com/77975803/224328764-dffffa72-7c8b-48d1-8435-8dd39751d132.png)

Then, just create a new page, type whatever you want and `save` the page more than 20 times. It will keep as maximum 20 revisions.

You can also test the migration by running `wp-env run cli wp p4-run-activator`.

## Testing (locally)
1. Make sure the field is already added into the `Settings > Writing` page.
2. Create a new page, add a dummy content and save it more than 20 times.
  a. If the `revisions_to_keep` is set you won't be able to save it more than 20 times.
3. Go back  `Settings > Writing` page and change the `revisions_to_keep` field to a lower value, then go back to the previous page.
4. Run the migration file by ` wp-env run cli wp p4-run-activator `.
5. If everything works as expected, you will see the value of `Revisions` of that page the same as the `revisions_to_keep` before updated.

Demo video

https://github.com/greenpeace/planet4-master-theme/assets/77975803/5b1609e2-a16a-4fa2-8d94-7af37b4bb6ac
